### PR TITLE
👔 Replace limiter lib with Postgres 

### PR DESCRIPTION
**Pull request by WeiJun0827**

### CODE REVIEW

File: src/commands/faucet.ts [hunk 1]
```diff
@@ -30,21 +31,33 @@ const FaucetModule = (cosmosClient: CosmosClient): Command => {
...
 
     try {
       const txHash = await cosmosClient.distribute(address);
-      rateLimiter.limit(address);
+      await Drop.create({
+        discordId: discordId,
+        address: address,
+        amount: Config.faucet.amount,
+        txHash: txHash,
+      });
```
Comment (WeiJun0827) [inline src/commands/faucet.ts]: Would we have race conditions here?

File: docker-compose.yml [hunk 2]
```diff
@@ -11,3 +11,10 @@ services:
     env_file:
       - ./.env
     command: ["yarn", "dev"]
+  db-dev:
+    image: postgres:12.3
+    profiles: [scripts]
```
Comment (williamchong) [inline docker-compose.yml]: What does this profile do?

File: src/commands/faucet.ts (hunk 1 above)
Comment (williamchong) [reply src/commands/faucet.ts]: Run `Drop.create` first, then try catch `cosmosClient.distribute`,  revert the drop if distribute  fail
Also we need `sequelize.transaction` between `Drop.findOne` and `Drop.create`

Comment (williamchong) [reply src/commands/faucet.ts]: What is the use of `User` table in this faucet btw, just for record?

**Review by williamchong — CHANGES_REQUESTED** [review]
Maybe we want to avoid address from claiming more than once too

File: src/commands/faucet.ts (hunk 1 above)
Comment (WeiJun0827) [reply src/commands/faucet.ts]: > What is the use of `User` table in this faucet btw, just for record?

Copied the DB code from [discord-bot](https://github.com/likecoin/likecoin-discord-bot/blob/main/backend/src/db.js). 
Probably can drop this `User` table in our scenario.

File: docker-compose.yml (hunk 2 above)
Comment (WeiJun0827) [reply docker-compose.yml]: We can run `docker compose up` without triggering the `db-dev` service. And run `docker compose run db-dev` to launch it separately. 
https://docs.docker.com/compose/profiles
Learned this trick from [chain indexer](https://github.com/likecoin/likecoin-chain-tx-indexer/blob/master/docker-compose.yml#L53).

File: src/commands/faucet.ts (hunk 1 above)
Comment (WeiJun0827) [reply src/commands/faucet.ts]: > Run `Drop.create` first, then try catch `cosmosClient.distribute`

but we don't have `txHash` yet before running `cosmosClient.distribute`?

Comment (williamchong) [reply src/commands/faucet.ts]: we need the lock though, maybe use "pending" then update it after done?

Comment (WeiJun0827) [reply src/commands/faucet.ts]: Used transaction and `lock` config. See if it's OK to just create a data row once.
(Since if we create a pending row first and then update `txHash` later, the transaction will still roll it back if tx is failed?)

ref: 
https://sequelize.org/docs/v6/other-topics/transactions/#locks
https://stackoverflow.com/a/70889313/7978205

File: src/commands/faucet.ts [hunk 3]
```diff
@@ -35,38 +35,37 @@ const FaucetModule = (cosmosClient: CosmosClient): Command => {
...
-      await Drop.create({
-        discordId: discordId,
-        address: address,
-        amount: Config.faucet.amount,
-        txHash: txHash,
-      });
-      await User.upsert({
-        username: interaction.user.username,
-        discordId: discordId,
+    const result = await sequelize.transaction(async (t) => {
```
Comment (williamchong) [inline src/commands/faucet.ts]: i think we need to set isolation level to serialisable since we are locking rows that are yet to be created

File: src/commands/faucet.ts [hunk 4]
```diff
@@ -30,30 +31,41 @@ const FaucetModule = (cosmosClient: CosmosClient): Command => {
...
-      );
-    } catch (err: unknown) {
-      console.error(`Failed to create transaction to ${address} = `, err);
-      await interaction.editReply(
-        `:warning: Failed to create transaction to \`${address}\`, Please try again later`,
-      );
-    }
+      try {
+        const txHash = await cosmosClient.distribute(address);
+        await Drop.create({
```
Comment (williamchong) [inline src/commands/faucet.ts]: i think i prefer to create the drop with placeholder tx hash first and commit and tx. Sending the tx might prolong the lock unnecessarily

**Review by williamchong — CHANGES_REQUESTED** [review]
need some input from @nnkken about pgsql tx and lock

File: docker-compose.yml (hunk 2 above)
Comment (williamchong) [reply docker-compose.yml]: if docker compose is for dev usage, we do want to spin up  a pgsql server when dev though

Comment (WeiJun0827) [reply docker-compose.yml]: Will we just reuse the discord-bot's pgsql server?

Comment (williamchong) [reply docker-compose.yml]: we don't really care if we will share pgsql in prod here since docker compose is only for dev

Comment (WeiJun0827) [reply docker-compose.yml]: I see, and is there any after-action that should be done for this thread?

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      "preLaunchTask": "tsc: 建置 - tsconfig.json",
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "outFiles": [
+        "${workspaceFolder}/dist/**/*.js",
+      ]
+    }
+  ]
+}
```
Comment (nnkken) [inline .vscode/launch.json:33]: Is this file needed?

File: .env.example [hunk 6]
```diff
@@ -1,11 +1,17 @@
 DISCORD_TOKEN=
 CHANNEL_ID=
 
-FAUCET_MNEMONIC=
+FAUCET_MNEMONIC=likecoin-public-testnet-5
```
Comment (nnkken) [inline .env.example]: Why `mnemonic` is `likecoin-public-testnet-5`?
If it's just an example / placeholder, please give an example with the expected format (e.g. for mnemonic it should be 12 / 24 words?)

File: .vscode/launch.json (hunk 5 above)
Comment (WeiJun0827) [reply .vscode/launch.json:33]: It could be stashed locally, but if someone would like to use debugger to develop someday, the provided `launch.json` would be handy. Could be removed if redundant.

File: .env.example (hunk 6 above)
Comment (WeiJun0827) [reply .env.example]: Umm.. typo, fixed

File: docker-compose.yml (hunk 2 above)
Comment (williamchong) [reply docker-compose.yml]: remove the profile, let docker compose spin up pgsql by default, and let the default config connect to this pgsql by default

Comment (WeiJun0827) [reply docker-compose.yml]: OK, updated

**Review by williamchong — APPROVED** [review]
